### PR TITLE
AffineBeta distribution

### PR DESCRIPTION
**Pull request by ordabayevy**
An implementation of the `AffineBeta` distribution.

**Distribution parameters**
There are other possible ways to parameterize `AffineBeta`, e.g. `low,high` instead of `loc,scale` or `mean,size` instead of `concentration1,concentration0`. Which approach is more preferred?

One way to achieve alternative parameterization is by subclassing `AffineBeta`:

<details>

```py
class NewAffineBeta(AffineBeta):
    """
    :param mean: mean of the distribution.
    :param size: size parameter of the Beta distribution.
    :param low: min parameter.
    :param high: max parameter.
    """

    arg_constraints = {
        "mean": constraints.dependent,
        "size": constraints.real,
        "low": constraints.real,
        "high": constraints.dependent,
    }

    def __init__(self, mean, size, low, high, validate_args=None):
        try:
            concentration1 = size * (mean - low) / (high - low)
            concentration0 = size * (high - mean) / (high - low)
        except ZeroDivisionError:
            # this is needed to work with funsor make_dist
            low = torch.tensor(0.0)
            high = torch.tensor(1.0)
            concentration1 = torch.tensor(1.0)
            concentration0 = torch.tensor(1.0)
        super(AffineBeta, self).__init__(
            concentration1,
            concentration0,
            low,
            high - low,
            validate_args=validate_args,
        )

    def expand(self, batch_shape, _instance=None):
        new = self._get_checked_instance(NewAffineBeta, _instance)
        return super(NewAffineBeta, self).expand(batch_shape, _instance=new)
```

</details>

**Clamping sample values**

To avoid  `NaN` and `Inf` values in the gradient sample values are clamped by  `torch.finfo(x.dtype).eps * self.scale`.

### CODE REVIEW

File: pyro/distributions/affine_beta.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    def concentration1(self):
+        return self.base_dist.concentration1
+
+    @property
+    def concentration0(self):
+        return self.base_dist.concentration0
+
+    @property
+    def size(self):
+        return self.concentration1 + self.concentration0
```
Comment (fritzo) [inline pyro/distributions/affine_beta.py:83]: This name of property seems a confusing, since in PyTorch `.size()` is often used as a discrete quantity. Would you mind omitting this one property, maybe defining it in your user code?

File: pyro/distributions/affine_beta.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    arg_constraints = {
+        "concentration1": constraints.positive,
+        "concentration0": constraints.positive,
+        "loc": constraints.real,
+        "scale": constraints.positive,
+    }
+
+    def __init__(self, concentration1, concentration0, loc, scale, validate_args=None):
+        base_dist = Beta(concentration1, concentration0)
```
Comment (fritzo) [inline pyro/distributions/affine_beta.py:31]:
```py
Beta(concentration1, concentration0, validate_args=validate_args)
```

File: pyro/distributions/affine_beta.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        Additionally clamps the output in order to avoid `NaN` and `Inf` values
+        in the gradients.
+        """
+        with torch.no_grad():
+            x = self.base_dist.sample(sample_shape)
+            for transform in self.transforms:
+                x = transform(x)
+            eps = torch.finfo(x.dtype).eps * self.scale
+            x = torch.min(torch.max(x, self.low + eps), self.high - eps)
+            return x
```
Comment (fritzo) [inline pyro/distributions/affine_beta.py:54]: I believe you can simply omit `.sample()`, since [Distribution.sample()](https://github.com/pytorch/pytorch/blob/0f412aa29392d75e33f7e37d8276be8b40b4ae73/torch/distributions/distribution.py#L123-L129) defaults to `torch.no_grad()(rsample)`.

**Review by fritzo — COMMENTED** [review]
Looks great!

File: pyro/distributions/affine_beta.py (hunk 1 above)
Comment (ordabayevy) [reply pyro/distributions/affine_beta.py:83]: If I rename it to `.sample_size` like here https://en.wikipedia.org/wiki/Beta_distribution#Mean_and_sample_size would it still be confusing? If that doesn't work either I'll remove it from here and just keep it in my code as you suggested.

File: tests/distributions/conftest.py [hunk 4]
```diff
@@ -131,6 +131,20 @@ def __init__(self, loc, scale):
...
+            scipy_dist=sp.beta,
+            examples=[
+                {'concentration1': [2.4], 'concentration0': [3.6], 'loc': [-1.0], 'scale': [2.0],
+                 'test_data': [-0.4]},
+                {'concentration1': [[2.4, 2.4], [3.6, 3.6]], 'concentration0': [[2.5, 2.5], [2.5, 2.5]],
+                 'loc': [[-1.0, -1.0], [2.0, 2.0]], 'scale': [[2.0, 2.0], [1.0, 1.0]],
+                 'test_data': [[[-0.4, 0.4], [2.5, 2.6]]]},
+                {'concentration1': [[2.4], [3.7]], 'concentration0': [[3.6], [2.5]],
+                 'loc': [[-1.0], [2.0]], 'scale': [[2.0], [2.0]],
+                 'test_data': [[0.0], [3.0]]}
```
Comment (ordabayevy) [inline tests/distributions/conftest.py:144]: Are there any criteria for setting `examples` values here? I just took values from `dist.Beta` and slightly changed them.

File: pyro/distributions/affine_beta.py (hunk 1 above)
Comment (fritzo) [reply pyro/distributions/affine_beta.py:83]: Sure, `.sample_size` seems ok.

File: tests/distributions/conftest.py (hunk 4 above)
Comment (fritzo) [reply tests/distributions/conftest.py:144]: There are no specific criteria. It's fine if you simply copy `Beta`. We just want to test some nontrivial values and some batched values.

File: pyro/distributions/affine_beta.py (hunk 3 above)
Comment (ordabayevy) [reply pyro/distributions/affine_beta.py]: `.sample()` is redefined in [`TransformedDistribution`](https://github.com/pytorch/pytorch/blob/0f412aa29392d75e33f7e37d8276be8b40b4ae73/torch/distributions/transformed_distribution.py#L78-L89). Maybe I should add it back.

Comment (fritzo) [reply pyro/distributions/affine_beta.py:54]: ok, this is fine then.

**Review by fritzo — APPROVED** [review]

**Comment by ordabayevy** [comment]
Thanks for reviewing it!
